messaging/src/main/java/io/lantern/messaging/CryptoWorker.kt

```diff
@@ -0,0 +1,837 @@
+                }
+            }
+
+            Model.OutboundMessage.ContentCase.DELETEMESSAGEID -> {
+                encryptAndSendToAll(out) {
+                    Model.TransferMessage.newBuilder().setDeleteMessageId(out.deleteMessageId)
+                        .build()
+                        .toByteArray()
+                }
+            }
+
+            Model.OutboundMessage.ContentCase.DISAPPEARSETTINGS -> {
+                encryptAndSendToAll(out) {
+                    Model.TransferMessage.newBuilder().setDisappearSettings(out.disappearSettings)
+                        .build()
+                        .toByteArray()
+                }
+            }
+
+            else -> logger.error("unknown outbound message content type")
+        }
+    }
+
+    private fun retrievePreKeys(out: Model.OutboundMessage.Builder) {
+        val recipientIdentityKey = out.recipientIdentityKey
+        messaging.anonymousClientWorker.withClient { client ->
+            client.requestPreKeys(
+                recipientIdentityKey,
+                emptyList(),
+                object : Callback<List<Messages.PreKey>> {
+                    override fun onSuccess(result: List<Messages.PreKey>) {
+                        submit {
+                            db.mutate {
+                                result.forEach { preKey ->
+                                    val oneTimePreKey = preKey.oneTimePreKey?.let {
+                                        // it's okay for oneTimePreKey to be empty
+                                        if (it.size() > 0) PreKeyRecord(it.toByteArray()) else null
+                                    }
+                                    val signedPreKey =
+                                        SignedPreKeyRecord(preKey.signedPreKey.toByteArray())
+                                    // TODO: implement max_recv checking for signed pre key age
+                                    val builder = SessionBuilder(
+                                        store,
+                                        SignalProtocolAddress(
+                                            recipientIdentityKey,
+                                            DeviceId(preKey.deviceId.toByteArray())
+                                        )
+                                    )
+                                    builder.process(
+                                        PreKeyBundle(
+                                            oneTimePreKey?.id ?: 0,
+                                            oneTimePreKey?.keyPair?.publicKey,
+                                            signedPreKey.id,
+                                            signedPreKey.keyPair.publicKey,
+                                            signedPreKey.signature,
+                                            recipientIdentityKey
+                                        )
+                                    )
+                                }
+                            }
```

Comment:
SessionBuilder has side effects that mutate the database. It will create its own transaction if none is in flight, but by wrapping this whole thing in a mutate we make the building of all Sessions atomic.